### DISCUSSION TITLE
Blend images with transparency mask

### DISCUSSION THREAD

**Discussion by SoupyGit** [ImageMagick/ImageMagick#5402] [Digital Image Processing] [answered]
I am trying to recreate the following process from Krita with ImageMagick.

In Krita:
a) Open OUTFILETIFF2 as bottom layer
b) Open OUTFILETIFF1 as top layer
c) Duplicate OUTFILETIFF1, convert to grayscale, auto levels, edge detection with radius 1. Convert this to a transparency mask applied to OUTFILETIFF1.
d) Export Jpeg.

This is my current attempt, with a little bash:

magick -quiet "${OUTFILETIFF1}"
  ( -clone 0 -colorspace gray -auto-level -canny 1 ) -compose over -composite -set colorspace sRGB +depth -compress none "${OUTFILETIFF2}" \
  -intent Perceptual -black-point-compensation -profile "${ICCPROFILE}" \
  -quality 92 \
  "${OUTFILE6}"
  
It is not outputting any image. Is anyone able to spot the error? Or recommend a better solution?
Much appreciated.

ImageMagick 7
Linux Mint 20.3

**Comment by snibgo** [comment]
> It is not outputting any image.

Are you sure? I suspect it is outputting two images. The first output would be the composition of OUTFILETIFF1 over itself. The second would be OUTFILETIFF2. Both outputs would be with the new profile.

You say you want "Convert this to a transparency mask applied to OUTFILETIFF1" but nothing in your command does this.

I suggest you show sample inputs, and the required output.

**Comment by SoupyGit** [comment]
> Are you sure? I suspect it is outputting two images.

Right. I should have said it is not outputting OUTFILE6.

> You say you want "Convert this to a transparency mask applied to OUTFILETIFF1" but nothing in your command does this.

What is the step I am missing?

> I suggest you show sample inputs, and the required output.

I have tried to upload a screenshot showing the Krita layers in .png, .jpg and .txt but it won't let me. All I can say is the layers look like this:

OUTFILETIFF1 (With Transparency Mask)
OUTFILETIFF2

Where OUTFILETIFF1 and OUTFILETIFF2 are inputs as per (a) and (b) above.
And Transparency Mask was created as per (c) above.
All that's left to do from there is export as jpg.

**Reply by fmw42** [reply]
zip your files. Then you can upload.

**Reply by SoupyGit** [reply]
Nope, still won't allow it. Tried as both .zip and .tar.gz

**Comment by fmw42** [comment]
I have no issues uploading zip files. How big is the zip file

**Reply by SoupyGit** [reply]
Only 176kb.

**Comment by snibgo** [comment]
I would split the OP command into lines like this:
```
magick -quiet \
  "${OUTFILETIFF1}" \
  ( -clone 0 -colorspace gray -auto-level -canny 1 ) \
  -compose over -composite \
  -set colorspace sRGB \
  +depth -compress none \
  "${OUTFILETIFF2}" \
  -intent Perceptual -black-point-compensation -profile "${ICCPROFILE}" \
  -quality 92 \
  "${OUTFILE6}"
```
We can see more clearly that you read OUTFILETIFF1, then make a clone of it so you then have to images in the list, then you composite those images together so we again have just one image in the list.

You set that image to sRGB, and set depth and compression.

Then you read OUTFILETIFF2, so you then have two images in the list.

Then you change or set the profile of both images, set the quality, and write them both. For example, if you have OUTFILE6 set to "myfile.jpg" then IM will write two files named "myfile-1.jpg" and "myfile-2.jpg".

If OUTFILETIFF2 represents the transparency you want in the main image, where black represents fully transparent and white represents fully opaque, then you should have:
```
-alpha off -compose CopyOpacity -composite
```
... after reading OUTFILETIFF2. This will composite the two images, so there is only one image in the list, so the final write will be to only one file, named "myfile.jpg" or whatever.

The "alpha off" will switch off any alpha that either image has. However, if OUTFILETIFF2 has transparency, and you want that transparency to be copied, then remove "-alpha off".

However, you say "OUTFILETIFF1 (With Transparency Mask)", and that confuses me. "CopyOpacity" will replace the transparency. Do you want to keep the transparency that is in OUTFILETIFF1? How do you want to combine that with transparency from OUTFILETIFF2?

Or perhaps you want to replace "-compose over -composite" by "-compose CopyOpacity -composite". Sorry, I am guessing.

**Comment by SoupyGit** [comment]
> However, you say "OUTFILETIFF1 (With Transparency Mask)", and that confuses me. "CopyOpacity" will replace the transparency. Do you want to keep the transparency that is in OUTFILETIFF1? How do you want to combine that with transparency from OUTFILETIFF2?

Ok, the transparency mask created as per (c) should look white around image edges, and black everywhere else.
When applied to OUTFILETIFF1, it means only the edges of OUTFILETIFF1 will be visible. For everywhere else, whats visible is the layer underneath, which is OUTFILETIFF2.
Eg. If OUTFILETIFF1 is very sharp, and OUTFILETIFF2 is the same image only blurry, we want to apply a mask to OUTFILETIFF1 so that only its edges are shown. That way we get the sharp edges of OUTFILETIFF1, and the blurry everything else of OUTFILETIFF2.

> Then you change or set the profile of both images, set the quality, and write them both. 

Ah OK. What I intended to do with those commands was composite the transparency of the clone onto OUTFILETIFF1, but I got it wrong.
I guess -composite (and maybe compose over?) should be used later when I wish to merge OUTFILETIFF1 and 2 into one.

Appreciate the help.

**Comment by snibgo** [comment, marked as answer]
Okay. Suppose your two input files are:

toes.jpg:
![toes](https://user-images.githubusercontent.com/33812368/182403179-0b0f024d-d86d-4df3-ab2c-ad5616f7ac64.jpg)

b.jpg:
![b](https://user-images.githubusercontent.com/33812368/182403234-b74acde1-4e37-4515-ac11-1b6a9321b607.jpg)

A basic command (bash syntax) might be:
```
magick \
  toes.jpg \
  \( -clone 0 -colorspace gray -auto-level -canny 0x1+10%+70% +write x1.png \) \
  -compose CopyOpacity -composite +write x2.png \
  b.jpg \
  -compose DstOver -composite \
  out.png
```
The operations "`+write x1.png`" and "`+write x2.png`" are for debugging. They have no effect on the output, and can be removed. But they are useful to understand what is happening.

x1.png is:
![x1](https://user-images.githubusercontent.com/33812368/182403275-af3bdc57-6564-4eca-b0b8-3382d3719586.png)

x2.png is:
![x2](https://user-images.githubusercontent.com/33812368/182403306-216d8cba-c20e-4a69-a69b-5b95b0dd1472.png)

out.png is:
![out](https://user-images.githubusercontent.com/33812368/182403355-cf2268ad-5f9a-4c8f-a4e7-acbd7cd95b40.png)

We can hardly see the sharpening effect. We can dilate (enlarge) the area that becomes transparent:
```
magick \
  toes.jpg \
  \( -clone 0 -colorspace gray -auto-level -canny 0x1+10%+70% -morphology dilate disk:4 +write x1b.png \) \
  -compose CopyOpacity -composite +write x2b.png \
  b.jpg \
  -compose DstOver -composite \
  outb.png
```
x1b.png is:
![x1b](https://user-images.githubusercontent.com/33812368/182403406-428f4a79-fd76-4d51-b6f0-4fa08426a07c.png)

x2b.png is:
![x2b](https://user-images.githubusercontent.com/33812368/182403449-9f177750-913f-4397-beb5-0e88bf2b28a5.png)

outb.png is:
![outb](https://user-images.githubusercontent.com/33812368/182403488-75c0538d-0f29-4b85-b7cf-7501d9fa6d86.png)

The sharpening effect is more clear.

Does that help?

**Reply by SoupyGit** [reply]
Brilliant, exactly what I was after! Thanks a lot!
